NOnion/Network/TorGuard.fs

```diff
@@ -13,14 +13,98 @@ open NOnion
 open NOnion.Cells
 open NOnion.Utility
 
+// RequireQualifiedAccess is needed to prevent collision with
+// Failure function that creates general exceptions
+[<RequireQualifiedAccess>]
+type internal GuardSendResult =
+    | Ok
+    | Failure of exn
+
+type internal GuardSendMessage =
+    {
+        CircuitId: uint16
+        CellToSend: ICell
+        ReplyChannel: AsyncReplyChannel<GuardSendResult>
+    }
+
 type TorGuard private (client: TcpClient, sslStream: SslStream) =
     let shutdownToken = new CancellationTokenSource()
 
     let mutable circuitsMap: Map<uint16, ITorCircuit> = Map.empty
     // Prevents two circuit setup happening at once (to prevent race condition on writing to CircuitIds list)
     let circuitSetupLock: obj = obj()
 
-    let sendLock = SemaphoreLocker()
+    let sendMailBoxProcessor(inbox: MailboxProcessor<GuardSendMessage>) =
+        async {
+            let innerSend circuitId (cellToSend: ICell) =
```

Comment:
no need to declare this nested function inside the async block, put it outside

Reply:
Fixed in c5c9b5e18c3ff4052145c3a160806b6bbae1a2b2.